server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/Monitor.java

```diff
@@ -98,54 +127,246 @@ private void monitor(ProcessRef processRef) {
     for (WatcherThread watcherThread : watcherThreads) {
-      while (watcherThread.isAlive()) {
-        try {
-          watcherThread.join();
-        } catch (InterruptedException ignored) {
-          // ignore, stop blocking
-        }
+      awaitTermination(watcherThread);
+    }
+    trace("all child processes done");
+    return hasRestartBeenRequested(watcherThreads);
+  }
+
+  private static boolean hasRestartBeenRequested(List<WatcherThread> watcherThreads) {
+    for (WatcherThread watcherThread : watcherThreads) {
+      if (watcherThread.isAskedForRestart()) {
+        trace("one child process requested restart");
+        return true;
       }
     }
+    trace("no child process requested restart");
+    return false;
   }
 
   /**
    * Blocks until all processes are terminated.
    */
   public void stop() {
-    stopAsync();
-    try {
-      terminator.join();
-    } catch (InterruptedException ignored) {
-      // stop blocking and exiting
-    }
-    // safeguard if TerminatorThread is buggy
-    lifecycle.tryToMoveTo(State.STOPPED);
+    trace("start hard stop async...");
+    stopAsync(State.HARD_STOPPING);
+    trace("await termination of terminator...");
+    awaitTermination(terminator);
+    cleanAfterTermination();
+    trace("exit...");
     systemExit.exit(0);
   }
 
+  private void cleanAfterTermination() {
+    trace("go to STOPPED...");
+    if (lifecycle.tryToMoveTo(State.STOPPED)) {
+      trace("await termination of restartWatcher and hardStopWatcher...");
+      // wait for restartWatcher and hardStopWatcher to cleanly stop
+      awaitTermination(restartWatcher, hardStopWatcher);
+      trace("restartWatcher done");
+      // removing shutdown hook to avoid called stop() unnecessarily unless already in shutdownHook
+      if (!systemExit.isInShutdownHook()) {
+        trace("removing shutdown hook...");
+        Runtime.getRuntime().removeShutdownHook(shutdownHook);
+      }
+      // cleanly close JavaLauncher
+      launcher.close();
+    }
+  }
+
   /**
    * Asks for processes termination and returns without blocking until termination.
+   * However, if a termination request is already under way (it's not supposed to happen, but, technically, it can occur),
+   * this call will be blocking until the previous request finishes.
    */
   public void stopAsync() {
-    if (lifecycle.tryToMoveTo(State.STOPPING)) {
-      terminator.setProcesses(processes);
+    stopAsync(State.STOPPING);
+  }
+
+  private void stopAsync(State stoppingState) {
+    if (lifecycle.tryToMoveTo(stoppingState)) {
       terminator.start();
     }
   }
 
+  public void restartAsync() {
+    if (lifecycle.tryToMoveTo(State.RESTARTING)) {
+      restartor = new RestartorThread();
+      restartor.start();
+    }
+  }
+
+  /**
+   * Runs every time a restart request is detected.
+   */
+  private class RestartorThread extends Thread {
+
+    private RestartorThread() {
+      super("Restartor " + (restartorInstanceCounter++));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this increment or decrement operator into a dedicated statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)